src/dagmc/tools/ray_fire_test.cpp

```diff
@@ -0,0 +1,598 @@
+  double u = 2 * denom * rand() - 1; // randomly samples from -1 to 1.
+  uvw[0] = sqrt(1 - u * u) * cos(theta);
+  uvw[1] = sqrt(1 - u * u) * sin(theta);
+  uvw[2] = u;
+
+}
+
+/* program global data, including settings with their defaults*/
+typedef struct { CartVect p; CartVect v; } ray_t;
+std::vector< ray_t > rays; // list of user-specified rays (given with -f flag)
+
+static CartVect ray_source(0, 0, 0);
+
+static double source_rad = 0;
+static int vol_index = 1;
+static int num_random_rays = 1000;
+static int randseed = 12345;
+static bool do_stat_report = false;
+static bool do_trv_stats   = false;
+static double location_az = 2.0 * PI;
+static double direction_az = location_az;
+static const char* pyfile = NULL;
+
+static int random_rays_missed = 0; // count of random rays that did not hit a surface
+
+/* Most of the argument handling code was stolen/adapted from MOAB/test/obb/obb_test.cpp */
+static void usage(const char* error, const char* opt, const char* name = "ray_fire_test") {
+  const char* default_message = "Invalid option";
+  if (opt && !error)
+    error = default_message;
+
+  std::ostream& str = error ? std::cerr : std::cout;
+  if (error) {
+    str << error;
+    if (opt)
+      str << ": " << opt;
+    str << std::endl;
+  }
+
+  str << "Usage: " << name << " [options] input_file" << std::endl;
+  str << "       " << name << " -h" << std::endl;
+
+  if (!error) {
+    str << "-h  print this help" << std::endl;
+    str << "-s  print OBB tree structural statistics" << std::endl;
+    str << "-S  track and print OBB tree traversal statistics" << std::endl;
+    str << "-i <int>   specify volume to upon which to test ray intersections (default 1)" << std::endl;
+    str << "-n <int>   specify number of random rays to fire (default 1000)" << std::endl;
+    str << "-c <x> <y> <z>  Specify center of of random ray generation (default origin)." << std::endl;
+    str << "-r <real>  random ray radius.  Random rays begin at this distance from the center." << std::endl;
+    str << "           if < 0, fire rays inward through the center" << std::endl;
+    str << "           if >= 0, fire randomly directed rays.  (default 0)" << std::endl;
+    str << "-f <x> <y> <z> <u> <v> <w>  Fire one given ray and report result." << std::endl;
+    str << "           (May be given multiple times.  -f implies -n 0)" << std::endl;
+    str << "-z <int>   seed the random number generator (default 12345)" << std::endl;
+    str << "-L <real>  if present, limit random ray Location to between +-<value> degrees" << std::endl;
+    str << "-D <real>  if present, limit random ray Direction to between +-<value> degrees" << std::endl;
+    str << "           (unused if random ray radius < 0)" << std::endl;
+    str << "-p <filename>  if present, save parameters and results to a python dictionary" << std::endl;
+  }
+
+  exit(error ? 1 : 0);
+}
+
+static const char* get_option(int& i, int argc, char* argv[]) {
```

Comment:
Giving Information: There are a number of argument parsing libraries that help eliminate some of this boilerplate [1][2].  Personally I'm a fan of argparse [3], however getopt also I know is good [4].

[1] https://attractivechaos.wordpress.com/2018/08/31/a-survey-of-argument-parsing-libraries-in-c-c/.
[2] https://stackoverflow.com/questions/865668/how-to-parse-command-line-arguments-in-c
[3] https://github.com/hbristow/argparse
[4] http://www.gnu.org/savannah-checkouts/gnu/libc/manual/html_node/Getopt.html#Getopt